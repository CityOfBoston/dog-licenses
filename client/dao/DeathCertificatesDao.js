// @flow

import DataLoader from 'dataloader';
import type { LoopbackGraphql } from '../loopback-graphql';
import type { DeathCertificate } from '../types';

import fetchDeathCertificates from '../queries/fetch-death-certificates';
import searchDeathCertificates from '../queries/search-death-certificates';
import type { DeathCertificateSearchResults } from '../types';

export type DeathCertificateCache = {[id: string]: DeathCertificate};

export default class DeathCertificatesDao {
  loopbackGraphql: LoopbackGraphql;
  loader: DataLoader<string, ?DeathCertificate>;

  constructor(loopbackGraphql: LoopbackGraphql) {
    this.loopbackGraphql = loopbackGraphql;

    // create new array shenanigans to get Flow to accept that we're not returning Errors
    this.loader = new DataLoader((ids) => fetchDeathCertificates(loopbackGraphql, ids).then((a) => a.map((i) => i)));
  }

  get(id: string): Promise<?DeathCertificate> {
    return this.loader.load(id);
  }

  async search(query: string, page: number): Promise<DeathCertificateSearchResults> {
    const results = await searchDeathCertificates(this.loopbackGraphql, query, page);

    results.results.forEach((cert) => {
      this.loader.prime(cert.id, cert);
    });

    return results;
  }
}
