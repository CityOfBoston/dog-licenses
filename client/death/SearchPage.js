// @flow

import React from 'react';
import Head from 'next/head';
import Link from 'next/link';
import Router from 'next/router';
import type { Context } from 'next';

import type { ClientDependencies } from '../page';
import type { DeathCertificate } from '../types';

import type Cart from '../store/Cart';
import Nav from '../common/Nav';
import Pagination from '../common/Pagination';
import type { DeathCertificateSearchResults } from '../types';

import { GRAY_100 } from '../common/style-constants';

export type InitialProps = {|
  query: string,
  results: ?DeathCertificateSearchResults,
|}

export type Props = {
  /* :: ...InitialProps, */
  cart: Cart,
}

type State = {
  query: string,
};

export default class IndexPage extends React.Component {
  props: Props;
  state: State;

  static async getInitialProps(ctx: Context<*>, { deathCertificatesDao }: ClientDependencies): Promise<InitialProps> {
    const { query } = ctx;

    let results = null;

    if (query.q) {
      results = await deathCertificatesDao.search(query.q, parseInt(query.page, 10) || 1);
    }

    return {
      query: query.q || '',
      results,
    };
  }

  constructor(props: Props) {
    super(props);

    const { query } = props;

    this.state = {
      query,
    };
  }

  handleQueryChange = (ev: SyntheticInputEvent) => {
    this.setState({ query: ev.target.value });
  };

  handleSubmit = (ev: SyntheticInputEvent) => {
    const { query } = this.state;

    ev.preventDefault();
    Router.push(`/death?q=${encodeURIComponent(query)}`);
  }

  render() {
    const { results, cart } = this.props;
    const { query } = this.state;

    return (
      <div>
        <Head>
          <title>Boston.gov — Death Certificates</title>
        </Head>

        <Nav cart={cart} link="checkout" />

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title">Death Certificates</h1>
          </div>

          <form className="sf sf--md" acceptCharset="UTF-8" method="get" action="/death" onSubmit={this.handleSubmit}>
            <input name="utf8" type="hidden" value="✓" />

            <div className="sf-i">
              <input type="text" name="q" id="q" value={query} onChange={this.handleQueryChange} placeholder="Search…" className="sf-i-f" autoComplete="off" />
              <button className="sf-i-b" type="submit">Search</button>
            </div>
          </form>
        </div>

        { results && this.renderResults(results) }
        <div />

      </div>
    );
  }

  renderResults(results: DeathCertificateSearchResults) {
    // we want the query that was searched for
    const { query } = this.props;

    return (
      <div>
        <div className="p-a300 b--w">
          <div className="t--sans tt-u" style={{ fontSize: 12 }}>
            Showing { results.results.length } of { results.resultCount.toLocaleString() } results for “{query}”
          </div>
        </div>

        { results.results.map(this.renderResult) }

        { results.resultCount > results.results.length && this.renderPagination(results) }

        <div className="p-a300">
          Not finding what you’re looking for? Try refining your search or <a href="https://www.boston.gov/departments/registry/how-get-death-certificate" style={{ fontStyle: 'italic' }}>request a death certificate</a>.
        </div>
      </div>
    );
  }

  renderResult({ firstName, lastName, age, deathYear, id, causeOfDeath }: DeathCertificate) {
    return (
      <Link key={id} href={`/death/certificate?id=${id}`} as={`/death/certificate/${id}`}>
        <a className={'p-a300 br br-t100 b--w result'}>
          <div style={{ fontWeight: 'bold' }}>{firstName} {lastName}</div>
          <div>Died: {deathYear} Age: {age}</div>
          <div>Cause of death: {causeOfDeath ? 'Known' : 'Unknown'}</div>
          <style jsx>{`
            .result {
              display: block;
              color: inherit;
              font-style: italic;
              border-color: ${GRAY_100};
            }
          `}</style>
        </a>
      </Link>
    );
  }

  renderPagination({ page, pageCount }: DeathCertificateSearchResults) {
    const { query } = this.props;
    const makeHref = (p: number) => `/death?q=${query}&page=${p}`;

    return <Pagination page={page} pageCount={pageCount} hrefFunc={makeHref} />;
  }
}
