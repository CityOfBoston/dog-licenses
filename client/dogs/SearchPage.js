// @flow

import React from 'react';
import Head from 'next/head';

export type InitialProps = {||};

export type Props = {
  ...InitialProps,
};

type State = {
  firstName: string,
  lastName: string,
  dogName: string,
  year: string,
};

export default class IndexPage extends React.Component {
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);

    this.state = {
      firstName: '',
      lastName: '',
      dogName: '',
      year: '',
    };
  }

  handleInputChange = (ev: SyntheticInputEvent) => {
    const target = ev.target;
    const value = target.value;
    const name = target.name;
    this.setState({ [name]: value });
  };

  render() {
    return (
      <div>
        <Head>
          <title>Boston.gov — Dog Licenses</title>
        </Head>

        <div className="p-a300">
          <div className="sh sh--b0">
            <h1 className="sh-title">Renew Your Dog License</h1>
          </div>

          <form
            className="sf sf--md"
            acceptCharset="UTF-8"
            method="get"
            action="/dogs/form">
            <input name="utf8" type="hidden" value="✓" />

            <div className="txt">
              <label htmlFor="text" className="txt-l">Owner First Name</label>
              <input
                id="text"
                name="firstName"
                type="text"
                value={this.state.firstName}
                onChange={this.handleInputChange}
                placeholder="Emily"
                className="txt-f"
                required
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Owner Last Name</label>
              <input
                id="text"
                type="text"
                name="lastName"
                value={this.state.lastName}
                onChange={this.handleInputChange}
                placeholder="Elizabeth"
                className="txt-f"
                required
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Dog name</label>
              <input
                id="text"
                type="text"
                name="dogName"
                value={this.state.dogName}
                onChange={this.handleInputChange}
                placeholder="Clifford"
                className="txt-f"
                required
              />
            </div>

            <div className="txt">
              <label htmlFor="text" className="txt-l">Year licensed</label>
              <input
                id="text"
                type="text"
                name="year"
                value={this.state.year}
                onChange={this.handleInputChange}
                placeholder="2016"
                className="txt-f"
                required
              />
            </div>

            <button className="btn" type="submit">Search for id</button>

          </form>
        </div>
        <div />
      </div>
    );
  }
}
