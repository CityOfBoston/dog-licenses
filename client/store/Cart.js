// @flow

import { observable, computed, action } from 'mobx';

import type { DeathCertificate } from '../types';

export class CartItem {
  @observable.ref cert: DeathCertificate;
  @observable quantity: number;
}

export default class Cart {
  @observable items = [];

  @computed get size(): number {
    return this.items.reduce((acc, item) => acc + item.quantity, 0);
  }

  @action add(cert: DeathCertificate, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === cert.id);
    if (existingItem) {
      existingItem.quantity += quantity;
    } else {
      const item = new CartItem();
      item.cert = cert;
      item.quantity = quantity;

      this.items.push(item);
    }
  }

  @action setQuantity(certId: string, quantity: number) {
    const existingItem = this.items.find((item) => item.cert.id === certId);
    if (existingItem) {
      existingItem.quantity = quantity;
    }
  }

  @action remove(certId: string) {
    const idx = this.items.findIndex(({ cert }) => cert.id === certId);
    if (idx !== -1) {
      this.items.splice(idx, 1);
    }
  }
}
