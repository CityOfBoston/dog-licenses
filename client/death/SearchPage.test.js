// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Router from 'next/router';

import type { DeathCertificate, DeathCertificateSearchResults } from '../types';
import Cart from '../store/Cart';
import DeathCertificatesDao from '../dao/DeathCertificatesDao';

import SearchPage from './SearchPage';
import type { InitialProps } from './SearchPage';

jest.mock('next/router');
jest.mock('../dao/DeathCertificatesDao');

const TEST_DEATH_CERTIFICATES: DeathCertificate[] = [
  {
    id: '000001',
    firstName: 'Logan',
    lastName: 'Howlett',
    birthYear: '1974',
    deathYear: '2014',
    causeOfDeath: 'Adamantium suffocation',
    age: '4 yrs. 2 mos. 10 dys',
  },
  {
    id: '000002',
    firstName: 'Bruce',
    lastName: 'Banner',
    birthYear: '1962',
    deathYear: '2016',
    causeOfDeath: 'Hawkeye',
    age: '4 yrs. 2 mos. 10 dys',
  },
];

const TEST_SEARCH_RESULTS: DeathCertificateSearchResults = {
  results: TEST_DEATH_CERTIFICATES,
  resultCount: 50,
  page: 0,
  pageSize: 5,
  pageCount: 10,
};

const renderFromInitialProps = async (query: {[key: string]: string}, dependencies: Object) => {
  const cart = new Cart();

  const initialProps: InitialProps = await SearchPage.getInitialProps((({
    query,
  }): any), dependencies);

  return renderer.create(<SearchPage cart={cart} {...initialProps} />);
};

describe('rendering', () => {
  let deathCertificatesDao;

  beforeEach(() => {
    deathCertificatesDao = new DeathCertificatesDao((null: any));
  });

  it('shows empty search box', async () => {
    expect((await renderFromInitialProps({}, { deathCertificatesDao })).toJSON()).toMatchSnapshot();
  });

  it('shows search results', async () => {
    deathCertificatesDao.search.mockReturnValue(TEST_SEARCH_RESULTS);

    expect((await renderFromInitialProps({ q: 'Monkey Joe' }, { deathCertificatesDao })).toJSON()).toMatchSnapshot();
    expect(deathCertificatesDao.search).toHaveBeenCalledWith('Monkey Joe', 1);
  });
});

describe('searching', () => {
  it('redirects to search for a query', () => {
    const cart = new Cart();
    const wrapper = shallow(<SearchPage cart={cart} query="" results={null} />);

    wrapper.find('input[name="q"]').simulate('change', { target: { value: 'Monkey Joe' } });
    wrapper.find('form[action="/death"]').simulate('submit', { preventDefault: jest.fn() });

    expect(Router.push).toHaveBeenCalledWith('/death?q=Monkey%20Joe');
  });
});
