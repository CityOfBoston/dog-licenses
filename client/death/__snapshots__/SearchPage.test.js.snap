// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering shows empty search box 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={3477795240}
  >
    <div
      className="nv-s-l bar"
      data-jsx={3477795240}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={3477795240}
        href="/death/checkout"
        onClick={[Function]}
      >
        Checkout
      </a>
      <a
        className="cart-link"
        data-jsx={3477795240}
        href="/death/checkout"
        onClick={[Function]}
      >
        0
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0"
    >
      <h1
        className="sh-title"
      >
        Death Certificates
      </h1>
    </div>
    <form
      acceptCharset="UTF-8"
      action="/death"
      className="sf sf--md"
      method="get"
      onSubmit={[Function]}
    >
      <input
        name="utf8"
        type="hidden"
        value="✓"
      />
      <div
        className="sf-i"
      >
        <input
          autoComplete="off"
          className="sf-i-f"
          id="q"
          name="q"
          onChange={[Function]}
          placeholder="Search…"
          type="text"
          value=""
        />
        <button
          className="sf-i-b"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div />
</div>
`;

exports[`rendering shows search results 1`] = `
<div>
  <nav
    className="nv-s nv-s--sticky"
    data-jsx={3477795240}
  >
    <div
      className="nv-s-l bar"
      data-jsx={3477795240}
    >
      <a
        className="nv-s-l-b back-link"
        data-jsx={3477795240}
        href="/death/checkout"
        onClick={[Function]}
      >
        Checkout
      </a>
      <a
        className="cart-link"
        data-jsx={3477795240}
        href="/death/checkout"
        onClick={[Function]}
      >
        0
      </a>
    </div>
  </nav>
  <div
    className="p-a300"
  >
    <div
      className="sh sh--b0"
    >
      <h1
        className="sh-title"
      >
        Death Certificates
      </h1>
    </div>
    <form
      acceptCharset="UTF-8"
      action="/death"
      className="sf sf--md"
      method="get"
      onSubmit={[Function]}
    >
      <input
        name="utf8"
        type="hidden"
        value="✓"
      />
      <div
        className="sf-i"
      >
        <input
          autoComplete="off"
          className="sf-i-f"
          id="q"
          name="q"
          onChange={[Function]}
          placeholder="Search…"
          type="text"
          value="Monkey Joe"
        />
        <button
          className="sf-i-b"
          type="submit"
        >
          Search
        </button>
      </div>
    </form>
  </div>
  <div>
    <div
      className="p-a300 b--w"
    >
      <div
        className="t--sans tt-u"
        style={
          Object {
            "fontSize": 12,
          }
        }
      >
        Showing 
        2
         results for “
        Monkey Joe
        ”
      </div>
    </div>
    <a
      className="p-a300 br br-t100 b--w result"
      data-jsx={1723197323}
      href="/death/certificate/000001"
      onClick={[Function]}
    >
      <div
        data-jsx={1723197323}
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Logan
         
        Howlett
      </div>
      <div
        data-jsx={1723197323}
      >
        Born: 
        1974
         – Died: 
        2014
      </div>
      <div
        data-jsx={1723197323}
      >
        Cause of death: 
        Known
      </div>
    </a>
    <a
      className="p-a300 br br-t100 b--w result"
      data-jsx={1723197323}
      href="/death/certificate/000002"
      onClick={[Function]}
    >
      <div
        data-jsx={1723197323}
        style={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        Bruce
         
        Banner
      </div>
      <div
        data-jsx={1723197323}
      >
        Born: 
        1962
         – Died: 
        2016
      </div>
      <div
        data-jsx={1723197323}
      >
        Cause of death: 
        Known
      </div>
    </a>
    <div
      className="p-a300"
    >
      Not finding what you’re looking for? Try refining your search or 
      <a
        href="https://www.boston.gov/departments/registry/how-get-death-certificate"
        style={
          Object {
            "fontStyle": "italic",
          }
        }
      >
        request a death certificate
      </a>
      .
    </div>
  </div>
  <div />
</div>
`;
